File: code/java/ch21/practice/Warmup05Reverse.java
package ch21.practice;

import java.util.*;

/**
 * Warmup 5: Reverse a Linked List
 * Chapter 21: Linked Lists — Pointers and Connections
 *
 * PROBLEM: Build LL from array, reverse it, return result as array.
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Warmup05Reverse {
    public static int[] solve(int[] arr) {
        // TODO: Replace this with your solution
        return new int[]{};
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) arr[i] = sc.nextInt();
        System.out.println(Arrays.toString(solve(arr)));
        sc.close();
    }
}

File: code/python/ch21/practice/warmup_05_reverse.py
"""
Warmup 5: Reverse a Linked List
=================================
Chapter 21: Linked Lists — Pointers and Connections

PROBLEM
-------
Given an array of integers, build a singly linked list from the array,
reverse the linked list, and return the values as an array.

INPUT FORMAT
------------
A single line of space-separated integers (may be empty).

OUTPUT FORMAT
-------------
A list of integers in reversed order.

CONSTRAINTS
-----------
- 0 <= len(arr) <= 10^4
- -10^9 <= arr[i] <= 10^9

EXAMPLES
--------
Input:
  1 2 3 4 5
Output: [5, 4, 3, 2, 1]

Input:
  (empty)
Output: []

HINT
----
Use the 3-pointer approach: prev, current, next_node. At each step,
save current.next, point current.next to prev, advance prev and current.
When current is None, prev is the new head.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def solve(arr: list[int]) -> list[int]:
    """Reverse the linked list built from arr, return as list."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line = input().strip()
    if line:
        arr = list(map(int, line.split()))
    else:
        arr = []
    print(solve(arr))
